hbase-client/src/main/java/org/apache/hadoop/hbase/client/ConnectionRegistryRpcStubHolder.java

```diff
@@ -0,0 +1,212 @@
+      conf.getLong(HConstants.HBASE_RPC_TIMEOUT_KEY, HConstants.DEFAULT_HBASE_RPC_TIMEOUT));
+  }
+
+  private ImmutableMap<ServerName, ClientMetaService.Interface> createStubs(RpcClient rpcClient,
+    Collection<ServerName> addrs) {
+    LOG.debug("Going to use new servers to create stubs: {}", addrs);
+    Preconditions.checkNotNull(addrs);
+    ImmutableMap.Builder<ServerName, ClientMetaService.Interface> builder =
+      ImmutableMap.builderWithExpectedSize(addrs.size());
+    for (ServerName masterAddr : addrs) {
+      builder.put(masterAddr,
+        ClientMetaService.newStub(rpcClient.createRpcChannel(masterAddr, user, rpcTimeoutMs)));
+    }
+    return builder.build();
+  }
+
+  private void createStubsAndComplete(String clusterId,
+    CompletableFuture<ImmutableMap<ServerName, ClientMetaService.Interface>> future) {
+    RpcClient c = RpcClientFactory.createClient(conf, clusterId);
+    ImmutableMap<ServerName, ClientMetaService.Interface> m = createStubs(c, bootstrapNodes);
+    rpcClient = c;
+    addr2Stub = m;
+    future.complete(m);
+  }
+
+  private void createStubs(List<ServerName> bootstrapServers, int index,
+    CompletableFuture<ImmutableMap<ServerName, ClientMetaService.Interface>> future) {
+    // use null cluster id here as we do not know the cluster id yet, we will fetch it through this
+    // rpc client and use it to create the final rpc client
+    RpcClient getConnectionRegistryRpcClient = RpcClientFactory.createClient(noAuthConf, null);
+    // user and rpcTimeout are both not important here, as we will not actually send any rpc calls
+    // out, only a preamble connection header, but if we pass null as user, there will be NPE in
+    // some code paths...
+    RpcChannel channel =
+      getConnectionRegistryRpcClient.createRpcChannel(bootstrapServers.get(index), user, 0);
+    ConnectionRegistryService.Interface stub = ConnectionRegistryService.newStub(channel);
+    HBaseRpcController controller = rpcControllerFactory.newController();
+    stub.getConnectionRegistry(controller, GetConnectionRegistryRequest.getDefaultInstance(),
+      new RpcCallback<GetConnectionRegistryResponse>() {
+
+        @Override
+        public void run(GetConnectionRegistryResponse resp) {
+          synchronized (ConnectionRegistryRpcStubHolder.this) {
+            addr2StubFuture = null;
+            if (controller.failed()) {
+              if (ConnectionUtils.isUnexpectedPreambleHeaderException(controller.getFailed())) {
+                // this means we have connected to an old server where it does not support passing
+                // cluster id through preamble connnection header, so we fallback to use null
+                // cluster id, which is the old behavior
+                LOG.debug("Failed to get connection registry info, should be an old server,"
+                  + " fallback to use null cluster id", controller.getFailed());
+                createStubsAndComplete(null, future);
```

Comment:
I wonder, do we want to try the next bootstrap server? We can fall through to old behavior if all bootstrap servers fail preamble?

Reply:
We could but this will make the code even more complicated, since we need to record all the failures, and have to consider what should be done if we got different exceptions from different servers...
And if a server returns us this exception, it means that the cluster is with old server, or at least still under upgrading, then we keep the old behavior is acceptable?